dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -0,0 +1,190 @@
+
+#include <cuda.h>
+#include <unistd.h>
+
+#include <string>
+#include <memory>
+
+#include "dali/core/error_handling.h"
+#include "dali/core/cuda_utils.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/data/tensor.h"
+#include "dali/operators/reader/loader/video/nvdecode/ColorSpace.h"
+
+namespace dali {
+namespace detail {
+int process_video_sequence(void *user_data, CUVIDEOFORMAT *video_format) {
+  return video_format->min_num_decode_surfaces;
+}
+
+int process_picture_decode(void *user_data, CUVIDPICPARAMS *picture_params) {
+  FramesDecoderGpu *frames_decoder = static_cast<FramesDecoderGpu*>(user_data);
+  if (frames_decoder->flush_) {
+    frames_decoder->flush_ = false;
+    frames_decoder->decode_success_ = false;
+    return 0;
+  }
+
+  frames_decoder->decode_success_ = false;
+
+  CUDA_CALL(cuvidDecodePicture(frames_decoder->nvdecode_state_->decoder, picture_params));
+
+  // Copy decoded frame to output
+  CUVIDPROCPARAMS videoProcessingParameters = {};
+  videoProcessingParameters.progressive_frame = !picture_params->field_pic_flag;
+  videoProcessingParameters.second_field = 1;
+  videoProcessingParameters.top_field_first = picture_params->bottom_field_flag ^ 1;
+  videoProcessingParameters.unpaired_field = 0;
+  videoProcessingParameters.output_stream = 0;
+
+  CUdeviceptr frame = 0;
+  unsigned int pitch = 0;
+
+  CUDA_CALL(cuvidMapVideoFrame(
+    frames_decoder->nvdecode_state_->decoder,
+    picture_params->CurrPicIdx,
+    &frame,
+    &pitch,
+    &videoProcessingParameters));
+
+  if (frames_decoder->current_copy_to_output_) {
+    CUDA_CALL(cudaDeviceSynchronize());
+    Nv12ToColor32(
+      reinterpret_cast<uint8_t *>(frame),
+      pitch,
+      frames_decoder->current_frame_output_,
+      frames_decoder->Width()* 3,
+      frames_decoder->Width(),
+      frames_decoder->Height(),
+      1);
+    CUDA_CALL(cudaDeviceSynchronize());
+  }
+
+  CUDA_CALL(cuStreamSynchronize(0));
+  CUDA_CALL(cuvidUnmapVideoFrame(frames_decoder->nvdecode_state_->decoder, frame));
+
+  frames_decoder->decode_success_ = true;
+  return 1;
+}
+}  // namespace detail
+
+FramesDecoderGpu::FramesDecoderGpu(const std::string &filename) :
+  FramesDecoder(filename) {
+    nvdecode_state_ = std::make_unique<NvDecodeState>();
+
+    // Create nv decoder
+    CUVIDDECODECREATEINFO decoder_info = {};
+    memset(&decoder_info, 0, sizeof(CUVIDDECODECREATEINFO));
+
+    decoder_info.bitDepthMinus8 = 0;
+    decoder_info.ChromaFormat = cudaVideoChromaFormat_420;
+    decoder_info.CodecType = cudaVideoCodec_MPEG4;
+    decoder_info.ulHeight = Height();
+    decoder_info.ulWidth = Width();
+    decoder_info.ulMaxHeight = Height();
+    decoder_info.ulMaxWidth = Width();
+    decoder_info.ulTargetHeight = Height();
+    decoder_info.ulTargetWidth = Width();
+    decoder_info.ulNumDecodeSurfaces = 4;
+    decoder_info.ulNumOutputSurfaces = 1;
+
+    CUDA_CALL(cuvidCreateDecoder(&nvdecode_state_->decoder, &decoder_info));
```

Comment:
How expensive is that? What if we need to handle 100 000 videos, would that eat up whole RAM/GPU memory?

Reply:
The architectural idea for this is to have one `FramesDecoderGpu` for multiple video files and call `Reconfigure`, when we change it.

Reply:
Understood.